### PR TITLE
Feat/version provenance

### PR DESCRIPTION
**Pull request by darscan**
- [x] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/dep-graph/blob/master/.github/CONTRIBUTING.md) rules
- [x] Reviewed by Snyk internal team

#### What does this PR do?

- preserves `versionProvenance` in legacy conversions
- enables TS strict mode

### CODE REVIEW

File: test/legacy/to-dep-tree.test.ts [hunk 1]
```diff
@@ -130,14 +156,14 @@ describe('graphToDepTree with a linux pkgManager', () => {
     test('missing repository alias', async () => {
       const depGraphData = helpers.loadFixture('os-deb-graph.json');
       const depGraph = depGraphLib.createFromJSON(depGraphData);
-      delete depGraph.pkgManager.repositories[0].alias;
+      delete depGraph.pkgManager.repositories!![0].alias;
```
Comment (gjvis) [inline test/legacy/to-dep-tree.test.ts]: what does the double `!!` do here?

**Comment by darscan** [comment]
true, done

File: src/legacy/index.ts [hunk 2]
```diff
@@ -86,12 +93,18 @@ async function buildGraph(
 
     depNodesIds.push(depNodeId);
 
-    builder.addPkgNode(depPkg, depNodeId);
+    const nodeInfo: types.NodeInfo = {};
+
+    if (dep.versionProvenance) {
+      nodeInfo.versionProvenance = dep.versionProvenance;
```
Comment (gjvis) [inline src/legacy/index.ts]: should we be worried about hanging on to a reference here? potential source of bugs...

File: src/legacy/index.ts [hunk 3]
```diff
@@ -103,7 +116,14 @@ async function buildGraph(
       version: depTree.version,
     };
     pkgNodeId = getNodeId(pkg.name, pkg.version, treeHash);
-    builder.addPkgNode(pkg, pkgNodeId);
+
+    const nodeInfo: types.NodeInfo = {};
+
+    if (depTree.versionProvenance) {
+      nodeInfo.versionProvenance = depTree.versionProvenance;
```
Comment (gjvis) [inline src/legacy/index.ts]: should we be worried about hanging on to a reference here? potential source of bugs...

File: src/legacy/index.ts [hunk 4]
```diff
@@ -214,9 +235,13 @@ function constructTargetOS(depGraph: types.DepGraph): { name: string; version: s
 async function buildSubtree(
     depGraph: types.DepGraphInternal, nodeId: string, eventLoopSpinner: EventLoopSpinner): Promise<DepTree> {
   const nodePkg = depGraph.getNodePkg(nodeId);
+  const nodeInfo = depGraph.getNode(nodeId);
   const depTree: DepTree = {};
   depTree.name = nodePkg.name;
   depTree.version = nodePkg.version;
+  if (nodeInfo.versionProvenance) {
+    depTree.versionProvenance = nodeInfo.versionProvenance;
```
Comment (gjvis) [inline src/legacy/index.ts]: should we be worried about hanging on to a reference here? potential source of bugs...

File: src/legacy/index.ts (hunk 2 above)
Comment (darscan) [reply src/legacy/index.ts]: not as far as I can see.. do you see something else?

**Comment by michael-go** [comment]
@gjvis @darscan updated the README a-bit more in: https://github.com/snyk/dep-graph/pull/18/commits/7c4a87af25b57064cd9952c9c865f9ed4618e8be

**Comment by michael-go** [comment]
oops thanks @darscan - fixed

**Pull request merged**